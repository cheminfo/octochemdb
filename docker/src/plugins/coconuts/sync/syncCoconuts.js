import md5 from 'md5';

import getLastDocumentImported from '../../../sync/http/utils/getLastDocumentImported.js';
import getLastFileSync from '../../../sync/http/utils/getLastFileSync.js';
import debugLibrary from '../../../utils/Debug.js';
import { getTaxonomiesForCoconuts } from '../../activesOrNaturals/utils/utilsTaxonomies/getTaxonomiesForCoconuts.js';

import { parseCoconuts } from './utils/parseCoconuts.js';

/**
 * @description Synchronize the coconuts collection from the coconut database
 * @param {*} connection MongoDB connection
 * @returns {Promise} returns coconuts collections
 */
export async function sync(connection) {
  const debug = debugLibrary('syncCoconuts');
  let options = {
    collectionSource: process.env.COCONUT_SOURCE,
    destinationLocal: `${process.env.ORIGINAL_DATA_PATH}/coconuts/full`,
    collectionName: 'coconuts',
    filenameNew: 'coconuts',
    extensionNew: 'zip',
  };
  try {
    let sources;
    let lastFile;
    if (process.env.NODE_ENV === 'test') {
      lastFile = `${process.env.COCONUTS_SOURCE_TEST}`;
      sources = [lastFile];
    } else {
      // Get lastFile available in the online database, the local database collection and the progress of the import
      const lastFile = await getLastFileSync(options);
      sources = [lastFile.replace(`${process.env.ORIGINAL_DATA_PATH}`, '')];
    }
    const progress = await connection.getProgress(options.collectionName);
    let isTimeToUpdate = false;
    if (
      progress.dateEnd !== 0 &&
      Date.now() - progress.dateEnd >
        Number(process.env.COCONUT_UPDATE_INTERVAL) * 24 * 60 * 60 * 1000 &&
      md5(JSON.stringify(sources)) !== progress.sources
    ) {
      progress.dateStart = Date.now();
      await connection.setProgress(progress);
      isTimeToUpdate = true;
    }

    // Get last document imported
    const lastDocumentImported = await getLastDocumentImported(
      connection,
      progress,
      options.collectionName,
    );
    // Define counters
    let counter = 0;
    let imported = 0;
    let start = Date.now();

    // check if importation is necessary
    if (
      lastDocumentImported === null ||
      ((md5(JSON.stringify(sources)) !== progress.sources ||
        progress.state !== 'updated') &&
        isTimeToUpdate)
    ) {
      // define file to use for importation inside the zip file
      let fileName = 'uniqueNaturalProduct.bson';
      debug(`Start parsing: ${fileName}`);
      debug(lastFile);
      const collection = await connection.getCollection(options.collectionName);
      const logs = await connection.getImportationLog({
        collectionName: options.collectionName,
        sources,
        startSequenceID: progress.seq,
      });
      // Get taxonomies collection
      const collectionTaxonomies = await connection.getCollection('taxonomies');
      // create temporary collection to import
      const temporaryCollection = await connection.getCollection(
        `${options.collectionName}_tmp`,
      );
      // set progress to updating
      progress.state = 'updating';
      await connection.setProgress(progress);
      // start importation
      for await (const entry of parseCoconuts(lastFile, fileName, connection)) {
        counter++;
        if (process.env.TEST === 'true' && counter > 20) break;

        if (Date.now() - start > Number(process.env.DEBUG_THROTTLING)) {
          debug(`Processing: counter: ${counter} - imported: ${imported}`);
          start = Date.now();
        }
        /// Normalize Taxonomies
        if (entry.data.taxonomies) {
          let taxonomies = await getTaxonomiesForCoconuts(
            entry,
            collectionTaxonomies,
          );
          entry.data.taxonomies = taxonomies;
        }
        entry._seq = ++progress.seq;
        await temporaryCollection.updateOne(
          { _id: entry._id },
          { $set: entry },
          { upsert: true },
        );

        imported++;
      }
      // rename temporary collection to real collection
      await temporaryCollection.rename(options.collectionName, {
        dropTarget: true,
      });
      // set logs and progress to updated
      logs.dateEnd = Date.now();
      logs.endSequenceID = progress.seq;
      logs.status = 'updated';
      await connection.updateImportationLog(logs);
      progress.sources = md5(JSON.stringify(sources));
      progress.dateEnd = Date.now();
      progress.state = 'updated';
      // create indexes on the collection for faster search
      await connection.setProgress(progress);
      await collection.createIndex({ 'data.ocl.noStereoTautomerID': 1 });
      debug(`${imported} compounds processed`);
    } else {
      debug(`file already processed`);
    }
  } catch (e) {
    if (connection) {
      debug(e.message, {
        collection: options.collectionName,
        connection,
        stack: e.stack,
      });
    }
  }
}
