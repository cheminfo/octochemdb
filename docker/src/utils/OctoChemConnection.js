import delay from 'delay';
import dotenv from 'dotenv';
import md5 from 'md5';
import { MongoClient } from 'mongodb';

import debugLibrary from './Debug.js';

dotenv.config();
const debug = debugLibrary('OctoChemConnection');

export function OctoChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

OctoChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

OctoChemConnection.prototype.getCollectionNames =
  async function getCollectionNames() {
    return (await (await this.getDatabase()).listCollections().toArray()).map(
      (entry) => entry.name,
    );
  };

OctoChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

OctoChemConnection.prototype.getImportationLog =
  async function getImportationLog(options) {
    const { collectionName, sources, startSequenceID } = options;
    const logsCollection = await this.getImportationLogsCollection();
    const sourcesHash = md5(JSON.stringify(sources));
    const _id = sourcesHash;
    let logs = await logsCollection.find({ _id }).next();
    if (logs === null) {
      logs = {
        _id,
        collectionName,
        sources,
        sourcesHash,
        dateStart: Date.now(),
        dateEnd: Date.now(),
        startSequenceID,
        endSequenceID: 0,
        status: 'updating',
      };
      await logsCollection.insertOne(logs);
    }
    return logs;
  };
OctoChemConnection.prototype.updateImportationLog =
  async function updateImportationLog(logs) {
    const collection = await this.getImportationLogsCollection();
    logs.dateEnd = Date.now();
    await collection.replaceOne({ _id: logs._id }, logs);
  };

OctoChemConnection.prototype.getImportationLogsCollection =
  async function getCollection() {
    return (await this.getDatabase()).collection('importationLogs');
  };

OctoChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

OctoChemConnection.prototype.setProgress = async function setProgress(
  progress,
) {
  const collection = await this.getAdminCollection();
  await collection.replaceOne({ _id: progress._id }, progress);
};

OctoChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    progress = {
      _id,
      state: 'updating',
      seq: 0,
      dateStart: Date.now(),
      dateEnd: 0,
    };
    await adminCollection.insertOne(progress);
  }
  return progress;
};

OctoChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      debug(e.stack);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

OctoChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};
