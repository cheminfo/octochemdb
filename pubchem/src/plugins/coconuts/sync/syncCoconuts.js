import md5 from 'md5';

import getLastDocumentImported from '../../../sync/http/utils/getLastDocumentImported.js';
import getLastFileSync from '../../../sync/http/utils/getLastFileSync.js';
import Debug from '../../../utils/Debug.js';
import { getTaxonomiesForCoconuts } from '../../activesOrNaturals/utils/utilsTaxonomies/getTaxonomiesForCoconuts.js';

import { parseCoconuts } from './utils/parseCoconuts.js';

/**
 * @description Synchronize the coconuts collection from the coconut database
 * @param {*} connection MongoDB connection
 * @returns {Promise} returns coconuts collections
 */
export async function sync(connection) {
  const debug = Debug('syncCoconuts');
  let options = {
    collectionSource: process.env.COCONUT_SOURCE,
    destinationLocal: `${process.env.ORIGINAL_DATA_PATH}/coconuts/full`,
    collectionName: 'coconuts',
    filenameNew: 'coconuts',
    extensionNew: 'zip',
  };
  try {
    // Get lastFile available in the online database, the local database collection and the progress of the import
    const lastFile = await getLastFileSync(options);
    const sources = [lastFile.replace(process.env.ORIGINAL_DATA_PATH, '')];
    const progress = await connection.getProgress(options.collectionName);
    const collection = await connection.getCollection(options.collectionName);
    const logs = await connection.getImportationLog({
      collectionName: options.collectionName,
      sources,
      startSequenceID: progress.seq,
    });
    // Get taxonomies collection
    const collectionTaxonomies = await connection.getCollection('taxonomies');
    // Get last document imported
    const lastDocumentImported = await getLastDocumentImported(
      connection,
      progress,
      options.collectionName,
    );
    // Define counters
    let counter = 0;
    let imported = 0;
    let start = Date.now();
    // define file to use for importation inside the zip file
    let fileName = 'uniqueNaturalProduct.bson';
    // check if importation is necessary
    if (
      lastDocumentImported === null ||
      md5(JSON.stringify(sources)) !== progress.sources ||
      progress.state !== 'updated'
    ) {
      debug(`Start parsing: ${fileName}`);
      // create temporary collection to import
      const temporaryCollection = await connection.getCollection(
        'temporaryCoconuts',
      );
      // set progress to updating
      progress.state = 'updating';
      await connection.setProgress(progress);
      // start importation
      for await (const entry of parseCoconuts(lastFile, fileName, connection)) {
        counter++;
        if (process.env.TEST === 'true' && counter > 20) break;

        if (
          Date.now() - start >
          Number(process.env.DEBUG_THROTTLING || 10000)
        ) {
          debug(`Processing: counter: ${counter} - imported: ${imported}`);
          start = Date.now();
        }
        /// Normalize Taxonomies
        if (entry.data.taxonomies) {
          let taxonomies = await getTaxonomiesForCoconuts(
            entry,
            collectionTaxonomies,
          );
          entry.data.taxonomies = taxonomies;
        }
        entry._seq = ++progress.seq;
        await temporaryCollection.updateOne(
          { _id: entry._id },
          { $set: entry },
          { upsert: true },
        );

        imported++;
      }
      // rename temporary collection to real collection
      await temporaryCollection.rename(options.collectionName, {
        dropTarget: true,
      });
      // set logs and progress to updated
      logs.dateEnd = Date.now();
      logs.endSequenceID = progress.seq;
      logs.status = 'updated';
      await connection.updateImportationLog(logs);
      progress.sources = md5(JSON.stringify(sources));
      progress.dateEnd = Date.now();
      progress.state = 'updated';
      // create indexes on the collection for faster search
      await connection.setProgress(progress);
      await collection.createIndex({ _id: 1 });
      await collection.createIndex({ _seq: 1 });
      await collection.createIndex({ 'data.ocl.noStereoID': 1 });

      debug(`${imported} compounds processed`);
    } else {
      debug(`file already processed`);
    }
  } catch (e) {
    if (connection) {
      debug(e, { collection: options.collectionName, connection });
    }
  }
}
