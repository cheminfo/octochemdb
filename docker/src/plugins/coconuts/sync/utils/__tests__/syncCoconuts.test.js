

import { test, expect } from 'vitest';
import { OctoChemConnection } from '../../../../../utils/OctoChemConnection.js';
import { sync } from '../../syncCoconuts';

test('syncCoconuts', async () => {

  const connection = new OctoChemConnection();
  const taxonomiesCollection = await connection.getCollection('taxonomies');
  while (true) {
    if ((await taxonomiesCollection.countDocuments()) === 20) {
      break;
    }
  }
  await sync(connection);
  const collection = await connection.getCollection('coconuts');
  const collectionEntry = await collection.find({_id:'CNP0214016.1'})
  const result = await collectionEntry.toArray();

  expect(result).toMatchSnapshot();

  await connection.close();
});
