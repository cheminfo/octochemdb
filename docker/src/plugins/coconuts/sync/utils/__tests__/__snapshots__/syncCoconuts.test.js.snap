// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`syncCoconuts 1`] = `
[
  {
    "_id": "CNP0214016.1",
    "_seq": 2,
    "data": {
      "cas": "35846-53-8",
      "iupacName": "[(1~{S},2~{R},3~{S},5~{S},6~{S},16~{E},18~{E},20~{R},21~{S})-11-chloro-21-hydroxy-12,20-dimethoxy-2,5,9,16-tetramethyl-8,23-dioxo-4,24-dioxa-9,22-diazatetracyclo[19.3.1.1^{10,14}.0^{3,5}]hexacosa-10,12,14(26),16,18-pentaen-6-yl] (2~{S})-2-[acetyl(methyl)amino]propanoate",
      "name": "MAYTANSINE",
      "ocl": {
        "coordinates": "!B]T\\wDPJ[_ZQuQpBn]T\\wDPJ[JNgnbBwOZibDgWDwgBlR^}\\iAAHZYu[W?haxsSzwMNrXEWlRAAHEEsSc?h\`D[^LWNOUUIpCQAAI{D\`iGMvxwLWv[Bd_~b@Qm_YlJQp",
        "idCode": "e\`||EH@FOmRKLnml^US_PH\`hRDaHQD\`dcLRDaLRHQHsLaHRDcHqHRLGbShXdWUvBHbijZjjVjjjjjZRhPIEHZGRzORFDi@Jf\`",
        "noStereoTautomerID": "e\`||EH@FOmRKLnml^US_PH\`hRDaHQD\`dcLRDaLRHQHsLaHRDcHqHRLGbShXdWUvBHbjjZjjZjjjjjjbh@CTPsEfEl[Xup[awB^OtHrricxX~FvjX~FOacZfOacx[vicxX~F@",
      },
      "taxonomies": [
        {
          "dbRef": {
            "$id": "CNP0214016.1",
            "$ref": "coconuts",
          },
          "species": "Maytenus buchananii",
        },
        {
          "dbRef": {
            "$id": "CNP0214016.1",
            "$ref": "coconuts",
          },
          "species": "Maytenus confertiflorus",
        },
        {
          "dbRef": {
            "$id": "CNP0214016.1",
            "$ref": "coconuts",
          },
          "species": "Maytenus guangsinensis",
        },
        {
          "dbRef": {
            "$id": "CNP0214016.1",
            "$ref": "coconuts",
          },
          "species": "Maytenus hookeri",
        },
        {
          "dbRef": {
            "$id": "CNP0214016.1",
            "$ref": "coconuts",
          },
          "species": "Maytenus ovatus",
        },
        {
          "dbRef": {
            "$id": "CNP0214016.1",
            "$ref": "coconuts",
          },
          "species": "Maytenus senata",
        },
        {
          "dbRef": {
            "$id": "CNP0214016.1",
            "$ref": "coconuts",
          },
          "species": "Maytenus serrata",
        },
        {
          "dbRef": {
            "$id": "CNP0214016.1",
            "$ref": "coconuts",
          },
          "species": "Putterlickia verrucosa",
        },
        {
          "dbRef": {
            "$id": "CNP0214016.1",
            "$ref": "coconuts",
          },
          "species": "Candidatus Borrarchaeia",
        },
      ],
    },
  },
]
`;
